File: scripts/maintainance/delete_stale_branches.py
import os
from datetime import datetime, timedelta, UTC
import requests
import json
import semver  # Added semver library

from scripts.util.load_env import load_github_env_vars

# Load environment variables
ENV = load_github_env_vars()
OWNER = ENV.get("REPO_OWNER", "vectordotdev")
REPO = ENV.get("REPO_NAME", "vector")
TOKEN = ENV.get("GITHUB_TOKEN")

# GitHub API URL for branches
API_URL = f"https://api.github.com/repos/{OWNER}/{REPO}/branches"

# Set the headers for authentication
COMMON_HEADERS = {
    "Authorization": f"token {TOKEN}",
    "Accept": "application/vnd.github.v3+json"
}


def is_semver_branch(branch_name):
    """Check if the branch name is a valid SemVer string starting with 'v'."""
    if not branch_name.startswith('v'):
        return False
    try:
        # Remove the 'v' prefix and parse the rest as a SemVer string
        semver.VersionInfo.parse(branch_name[1:])
        return True
    except ValueError:
        return False


def get_last_commit_date(github_token, repo_owner, repo_name, branch_name):
    """Fetch the last commit date for the given branch from GitHub API"""
    url = f"https://api.github.com/repos/{repo_owner}/{repo_name}/commits?sha={branch_name}&per_page=1"
    headers = {"Authorization": f"token {github_token}", "Accept": "application/vnd.github.v3+json"}

    try:
        response = requests.get(url, headers=headers)
        response.raise_for_status()

        commits = response.json()
        if not commits:
            return None

        last_commit_date = commits[0]['commit']['committer']['date']
        return datetime.strptime(last_commit_date, "%Y-%m-%dT%H:%M:%SZ")

    except requests.exceptions.RequestException as e:
        print(f"Error fetching commit date for branch {branch_name}: {str(e)}")
        return None


def check_branch_activity(last_commit_date, activity_limit_years=4):
    """Check if the branch was updated within the activity limit (default 4 years)"""
    if last_commit_date is None:
        return False

    now = datetime.now(UTC)
    activity_limit = timedelta(days=activity_limit_years * 365)
    return (now - last_commit_date) <= activity_limit


def get_branches(page=1, per_page=100):
    """Fetches a list of all branches from the repository with pagination."""
    params = {"page": page, "per_page": per_page}
    try:
        response = requests.get(API_URL, headers=COMMON_HEADERS, params=params)
        response.raise_for_status()

        branches = response.json()
        link_header = response.headers.get("Link", "")
        has_next = "rel=\"next\"" in link_header
        return branches, has_next
    except requests.exceptions.RequestException as e:
        print(f"Error fetching branches (page {page}): {str(e)}")
        return [], False


def delete_branch(branch_name):
    """Deletes the specified branch using the GitHub API."""
    delete_url = f"https://api.github.com/repos/{OWNER}/{REPO}/git/refs/heads/{branch_name}"
    try:
        response = requests.delete(delete_url, headers=COMMON_HEADERS)
        if response.status_code == 204:
            print(f"Successfully deleted branch: {branch_name}")
        else:
            print(f"Failed to delete branch {branch_name}: {response.status_code} - {response.text}")
    except requests.exceptions.RequestException as e:
        print(f"Error deleting branch {branch_name}: {str(e)}")


def main():
    """Main function to list and delete stale branches with pagination."""
    if not all([OWNER, REPO, TOKEN]):
        print("Error: Missing required environment variables (REPO_OWNER, REPO_NAME, GITHUB_TOKEN)")
        return

    page = 1
    per_page = 100
    all_branches = []

    while True:
        branches, has_next = get_branches(page, per_page)
        if not branches:
            break
        all_branches.extend(branches)
        if not has_next:
            break
        page += 1

    if not all_branches:
        print("No branches found or failed to fetch branches.")
        return

    for branch in all_branches:
        branch_name = branch['name']

        # Skip protected branches, main/master, and valid SemVer branches
        if (branch.get('protected', False) or
                branch_name in ['main', 'master'] or
                is_semver_branch(branch_name)):
            print(f"Skipping special branch: {branch_name}")
            continue

        # Check branch activity
        last_commit_date = get_last_commit_date(TOKEN, OWNER, REPO, branch_name)

        if last_commit_date:
            if check_branch_activity(last_commit_date):
                print(f"Keeping active branch: {branch_name} (Last commit: {last_commit_date})")
            else:
                print(f"Deleting stale branch: {branch_name} (Last commit: {last_commit_date})")
                delete_branch(branch_name)
        else:
            print(f"Could not determine activity for branch: {branch_name}")


if __name__ == "__main__":
    main()

File: src/commands/delete_stale_branches.rs
use crate::config::Config;
use anyhow::{Context, Result};
use chrono::{DateTime, Duration, Utc};
use reqwest::blocking::Client;
use serde_json::Value;

const STALE_YEARS: i64 = 4;

pub fn run(config: &Config, dry_run: bool, yes: bool) -> Result<()> {
    let client = Client::new();
    let cutoff = Utc::now() - Duration::days(STALE_YEARS * 365);

    println!(
        "Fetching branches for {}/{}...",
        config.org, config.repo
    );
    let branches = fetch_all_branches(&client, config)?;
    println!("Total branches fetched: {}", branches.len());

    // First pass: classify without deleting
    let mut stale: Vec<(String, DateTime<Utc>)> = Vec::new();

    for branch in &branches {
        let name = branch["name"].as_str().unwrap_or("");
        let protected = branch["protected"].as_bool().unwrap_or(false);

        if protected || name == "main" || name == "master" || is_semver_branch(name) {
            println!("Skipping special branch: {name}");
            continue;
        }

        match get_last_commit_date(&client, config, name)? {
            Some(last_commit) if last_commit > cutoff => {
                println!(
                    "Keeping active branch: {name} (last commit: {})",
                    last_commit.date_naive()
                );
            }
            Some(last_commit) => {
                println!(
                    "Stale branch: {name} (last commit: {})",
                    last_commit.date_naive()
                );
                stale.push((name.to_string(), last_commit));
            }
            None => {
                println!("Could not determine activity for branch: {name}");
            }
        }
    }

    if stale.is_empty() {
        println!("No stale branches found.");
        return Ok(());
    }

    println!("\nFound {} stale branch(es) to delete.", stale.len());

    if dry_run {
        println!("[dry-run] The following branches would be deleted:");
        for (name, last_commit) in &stale {
            println!("  {name} (last commit: {})", last_commit.date_naive());
        }
        return Ok(());
    }

    if !crate::confirm(&format!("Delete {} stale branch(es)?", stale.len()), yes) {
        println!("Aborted.");
        return Ok(());
    }

    for (name, _) in &stale {
        delete_branch(&client, config, name)?;
    }

    println!("Done. Deleted {} branch(es).", stale.len());
    Ok(())
}

fn is_semver_branch(name: &str) -> bool {
    let stripped = name.strip_prefix('v').unwrap_or(name);
    semver::Version::parse(stripped).is_ok()
}

fn fetch_all_branches(client: &Client, config: &Config) -> Result<Vec<Value>> {
    let url = format!(
        "https://api.github.com/repos/{}/{}/branches",
        config.org, config.repo
    );
    let mut branches = Vec::new();
    let mut page = 1u32;

    loop {
        let resp = client
            .get(&url)
            .header("Authorization", format!("Bearer {}", config.github_token))
            .header("Accept", "application/vnd.github.v3+json")
            .header("User-Agent", "github-tools")
            .query(&[("per_page", "100"), ("page", &page.to_string())])
            .send()
            .context("Failed to fetch branches")?
            .error_for_status()?;

        let has_next = resp
            .headers()
            .get("link")
            .and_then(|v| v.to_str().ok())
            .map(|l| l.contains("rel=\"next\""))
            .unwrap_or(false);

        let batch: Vec<Value> = resp.json()?;
        if batch.is_empty() {
            break;
        }
        branches.extend(batch);
        if !has_next {
            break;
        }
        page += 1;
    }
    Ok(branches)
}

fn get_last_commit_date(
    client: &Client,
    config: &Config,
    branch: &str,
) -> Result<Option<DateTime<Utc>>> {
    let url = format!(
        "https://api.github.com/repos/{}/{}/commits",
        config.org, config.repo
    );
    let resp = client
        .get(&url)
        .header("Authorization", format!("Bearer {}", config.github_token))
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "github-tools")
        .query(&[("sha", branch), ("per_page", "1")])
        .send()?
        .error_for_status()?;

    let commits: Vec<Value> = resp.json()?;
    let date_str = commits
        .first()
        .and_then(|c| c["commit"]["committer"]["date"].as_str());

    Ok(date_str.and_then(|s| s.parse::<DateTime<Utc>>().ok()))
}

fn delete_branch(client: &Client, config: &Config, branch: &str) -> Result<()> {
    let url = format!(
        "https://api.github.com/repos/{}/{}/git/refs/heads/{}",
        config.org, config.repo, branch
    );
    let resp = client
        .delete(&url)
        .header("Authorization", format!("Bearer {}", config.github_token))
        .header("Accept", "application/vnd.github.v3+json")
        .header("User-Agent", "github-tools")
        .send()?;

    if resp.status().as_u16() == 204 {
        println!("Deleted branch: {branch}");
    } else {
        eprintln!("Failed to delete branch {branch}: {}", resp.status());
    }
    Ok(())
}
